src/input-states.js

```diff
@@ -0,0 +1,242 @@
+        return getPortStateBySocketId(node, vnid, portName, socketId);
+    else if (_.isObject(arguments[2]))
+        return setAllPortStates(node, vnid, arguments[2]);
+    else return getAllPortStates(node, vnid);
+}
+
+/**
+ * Returns a hash of port names to port states or array of states if the port is
+ * addressable/multi. This provides the caller with all portNames, distingishes between
+ * addressable/multi and non-addressable/multi ports, and provides the caller with all their
+ * states that will be useful when calling an updater.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ */
+function getAllPortStates(node, vnid) {
+    var inPorts = _.pick(node.inPorts, isPort);
+    return _.mapObject(inPorts, function(port, portName){
+        return getPortState(node, vnid, portName);
+    });
+}
+
+/**
+ * Returns the state of the port or an array of states, if the port is
+ * addressable/multi. When all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the InPort name on this node 
+ */
+function getPortState(node, vnid, portName) {
+    var port = node.inPorts[portName];
+    if (port.isMulti()) return getPortStateArray(node, vnid, portName);
+    else if (ensureVniStateExists(node, vnid, portName))
+        return node.vnis[vnid].inputStates[portName];
+    // By default states that come in on vnid '' apply to all VNIs
+    else if (vnid) return getPortState(node, '', portName);
+    else return undefined;
+}
+
+/**
+ * Returns an array of the port states, one item for each attached socket. When
+ * all the port states are not need this is a convient way to
+ * retrieve just the port state that is needed.
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the InPort name on this node 
+ */
+function getPortStateArray(node, vnid, portName) {
+    var port = node.inPorts[portName];
+    if (!isMultiPort(port)) throw Error("This port is not addressable/multi: " + portName);
+    else return port.listAttached().map(function(socketId){
+        return getPortStateBySocketId(node, vnid, portName, socketId);
+    });
+}
+
+/**
+ * Returns the state of this port for the given socketId. This can be used to
+ * look up the last state when checking to see if the state has changed.
+ * Note that the socketId is not neccessarily the same as the array index
+ * @param node a node facade
+ * @param vnid an identifier that distinguishes the set of VNI states
+ * @param portName the name of the addressable/multi port for this node
+ * @param socketId the socketId provided to a port data event
+ */
+function getPortStateBySocketId(node, vnid, portName, socketId) {
+    if (socketId == null)
+        return getPortState(node, vnid, portName);
```

Comment:
I want to make it easy for the caller. If the port is not multi then socketId is undefined. Rather then requiring the caller to check for null socketId and have two different functions, I think it would be easier to do that here and just have one public function for both cases.